### PR TITLE
Allow TaskGroup's ChildTaskResult Type To Be Inferred

### PR DESCRIPTION
**Pull request by rlziii**
Allows `TaskGroup` and `ThrowingTaskGroup`'s `ChildTaskResult` generic to be inferred in `withTaskGroup(of:returning:body:)`/`withThrowingTaskGroup(of:returning:body:)` functions.
 

Resolves #NNNNN, fix apple/llvm-project#MMMMM.

See: https://github.com/swiftlang/swift-evolution/pull/2494

### CODE REVIEW

**Comment by ktoso** [comment]
**Description**: 
**Scope/Impact**: 
**Risk:** 
**Testing**: 
**Reviewed by**: @hborla

**Original PR:** 
**Radar:**

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Thanks for the PR and proposal!

Since we were talking off github about this I took the liberty of pushing a reworked test-case that will pass :)
Generally we prefer using the `expected-error{{}}` etc, along with the `-verify` for such tests.

I simplified the test a bit to mostly be about the "new" behaviors and how we error report the "bad" cases.

I think this looks good, now we need to get it through Swift Evolution so I tagged the PR as such. 
Thank you again!

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: test/Concurrency/async_task_groups_and_type_inference.swift [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+let _ = await withThrowingTaskGroup { group in
+  group.addTask { }
+  group.addTask { }
+}
+
+let _ = await withTaskGroup { group in
+  group.addTask { 1 }
+  group.addTask { 2 }
+  return "result"
+}
```
Comment (rlziii) [inline test/Concurrency/async_task_groups_and_type_inference.swift:21]: @ktoso, should I also add a similar test as this for `withThrowingTaskGroup`?

**Comment by rlziii** [comment]
> Thanks for the PR and proposal!
> 
> Since we were talking off github about this I took the liberty of pushing a reworked test-case that will pass :) Generally we prefer using the `expected-error{{}}` etc, along with the `-verify` for such tests.
> 
> I simplified the test a bit to mostly be about the "new" behaviors and how we error report the "bad" cases.
> 
> I think this looks good, now we need to get it through Swift Evolution so I tagged the PR as such. Thank you again!

Thank you for fixing up the tests.  For the future, is there documentation for this testing syntax?

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

File: test/Concurrency/async_task_groups_and_type_inference.swift (hunk 1 above)
Comment (ktoso) [reply test/Concurrency/async_task_groups_and_type_inference.swift:21]: It wouldn't hurt, if you want to. Always good to cover all the APIs yes

**Comment by ktoso** [comment]
> For the future, is there documentation for this testing syntax?

There's some docs over here: https://github.com/swiftlang/swift/blob/main/docs/Diagnostics.md#diagnostic-verifier

Though generally looking at other existing tests is the best way really

File: test/Concurrency/async_task_groups_and_type_inference.swift (hunk 1 above)
Comment (rlziii) [reply test/Concurrency/async_task_groups_and_type_inference.swift:21]: I added the mirrored `withThrowingTaskGroup` test and also added a pair of tests to ensure that multi-statement closure inference works as expected, i.e.:

```swift
let _ = await withTaskGroup { group in
  group.addTask {
    if Bool.random() {
      return 1
    } else {
      return 0
    }
  }
  group.addTask {
    if Bool.random() {
      return 1
    } else {
      return 0
    }
  }
  return "result"
}
```

**Comment by rlziii** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

File: test/Concurrency/async_task_groups_and_type_inference.swift (hunk 1 above)
Comment (ktoso) [reply test/Concurrency/async_task_groups_and_type_inference.swift:21]: Thanks!

**Review by ktoso — APPROVED** [review]

**Comment by DougGregor** [comment]
@swift-ci please build toolchain Linux

**Comment by DougGregor** [comment]
SE-0422 has been [accepted](https://forums.swift.org/t/accepted-se-0422-allow-taskgroups-childtaskresult-type-to-be-inferred/73747).

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Pull request merged**